### ISSUE TITLE
MiniMax-H3: apply rope without the batched 2x2 matmul

### ISSUE THREAD

**Issue by sashaok123** [kohya-ss/musubi-tuner#1122]
On a 1280x704 fl2va run, the H3 training step spends more time in `_apply_rope_split_half` than in the attention it feeds.

`rotation_table` is built as `[B, S, 1, pairs, 2, 2]`, so `torch.matmul(rotation_table, rotary.unsqueeze(-1))` dispatches a batched GEMM whose batch is `B * S * H * pairs`. For a 209-frame 1280x704 sample that is 55768 packed rows * 56 heads * 48 pairs, roughly 150M products of 2x2 by 2x1, plus the `stack` and two `cat` temporaries around it. The useful arithmetic is a couple of GFLOP; nearly all of the cost is batch bookkeeping.

Writing the rotation out by hand is much cheaper, and with fp32 accumulation it is bit-identical: `torch.matmul` already accumulates the batched GEMM in fp32, so nothing changes numerically.

Measured on an H200 NVL, PyTorch 2.14.0+cu130, FlashAttention 3, `--convrot_int8`, `--base_weights` (training adapter v2), `--gradient_checkpointing`, dev at 70d42b8:

| | matmul | expanded |
| --- | --- | --- |
| rope, one tensor, 55768 rows | 113.2 ms | 6.1 ms |
| one DiTBlock, forward + backward | 1239 ms | 829 ms |
| peak allocated, one DiTBlock | 22.9 GiB | 20.7 GiB |
| training step, 12 steps, seed 42, identical data order | 82.44 s/it | 54.30 s/it |

Equivalence was checked on both forward and backward: `max|delta| = 0` in fp32 and in bf16. A pure bf16 expansion is another 2 ms faster but moves gradients by up to 0.125, so this keeps the fp32 accumulation.

After the change, the per-block CUDA profile is 66% FlashAttention, 22% cuBLAS, 8% everything else.

### Patch

```diff
 def _apply_rope_split_half(hidden_states: torch.Tensor, rotation_table: torch.Tensor) -> torch.Tensor:
     pairs = rotation_table.shape[-3]
-    rotary = torch.stack((hidden_states[..., :pairs], hidden_states[..., pairs : 2 * pairs]), dim=-1)
-    rotary = torch.matmul(rotation_table, rotary.unsqueeze(-1)).squeeze(-1)
-    rotary = torch.cat((rotary[..., 0], rotary[..., 1]), dim=-1)
-    return torch.cat((rotary, hidden_states[..., 2 * pairs :]), dim=-1)
+    cosine = rotation_table[..., 0, 0].float()
+    sine = rotation_table[..., 1, 0].float()
+    first = hidden_states[..., :pairs].float()
+    second = hidden_states[..., pairs : 2 * pairs].float()
+    dtype = hidden_states.dtype
+    return torch.cat(
+        (
+            (first * cosine - second * sine).to(dtype),
+            (first * sine + second * cosine).to(dtype),
+            hidden_states[..., 2 * pairs :],
+        ),
+        dim=-1,
+    )
```

### Reproduction

```python
import time
import torch

PAIRS, HEADS, HEAD_DIM, ROWS = 48, 56, 128, 55768


def matmul_form(hidden_states, rotation_table):
    pairs = rotation_table.shape[-3]
    rotary = torch.stack((hidden_states[..., :pairs], hidden_states[..., pairs : 2 * pairs]), dim=-1)
    rotary = torch.matmul(rotation_table, rotary.unsqueeze(-1)).squeeze(-1)
    rotary = torch.cat((rotary[..., 0], rotary[..., 1]), dim=-1)
    return torch.cat((rotary, hidden_states[..., 2 * pairs :]), dim=-1)


def expanded_form(hidden_states, rotation_table):
    pairs = rotation_table.shape[-3]
    cosine = rotation_table[..., 0, 0].float()
    sine = rotation_table[..., 1, 0].float()
    first = hidden_states[..., :pairs].float()
    second = hidden_states[..., pairs : 2 * pairs].float()
    dtype = hidden_states.dtype
    return torch.cat(
        (
            (first * cosine - second * sine).to(dtype),
            (first * sine + second * cosine).to(dtype),
            hidden_states[..., 2 * pairs :],
        ),
        dim=-1,
    )


def table(rows, dtype):
    angles = torch.rand(1, rows, 1, PAIRS, device="cuda") * 6.283185
    cosine, sine = torch.cos(angles), torch.sin(angles)
    return torch.stack((cosine, -sine, sine, cosine), dim=-1).reshape(1, rows, 1, PAIRS, 2, 2).to(dtype)


def timed(fn, *args, runs=3):
    for _ in range(2):
        fn(*args)
    torch.cuda.synchronize()
    start = time.perf_counter()
    for _ in range(runs):
        fn(*args)
    torch.cuda.synchronize()
    return (time.perf_counter() - start) / runs * 1000


for dtype in (torch.float32, torch.bfloat16):
    x = torch.randn(1, 8192, HEADS, HEAD_DIM, device="cuda", dtype=dtype, requires_grad=True)
    rot = table(8192, dtype)
    a, b = matmul_form(x, rot), expanded_form(x, rot)
    ga, = torch.autograd.grad(a.float().pow(2).sum(), x, retain_graph=True)
    gb, = torch.autograd.grad(b.float().pow(2).sum(), x)
    print(dtype, "fwd", (a.float() - b.float()).abs().max().item(), "bwd", (ga - gb).abs().max().item())

x = torch.randn(1, ROWS, HEADS, HEAD_DIM, device="cuda", dtype=torch.bfloat16)
rot = table(ROWS, torch.bfloat16)
print("matmul  ", timed(matmul_form, x, rot), "ms")
print("expanded", timed(expanded_form, x, rot), "ms")
```

I can open a pull request for this if you prefer.
